caluma/caluma_form/domain_logic.py

```diff
@@ -0,0 +1,78 @@
+from django.db import transaction
+
+from caluma.caluma_form import models, validators
+from caluma.caluma_user.models import BaseUser
+from caluma.utils import update_model
+
+
+class SaveAnswerLogic:
+    @staticmethod
+    def validate_for_save(
+        data: dict, user: BaseUser, answer: models.Answer = None, origin: bool = False
+    ) -> dict:
+
+        if data["question"].type == models.Question.TYPE_TABLE:
+            data["value"] = models.Document.objects.filter(pk__in=data["value"])
+
+        validators.AnswerValidator().validate(
+            **data, user=user, instance=answer, origin=origin
+        )
+
+        return data
+
+    @staticmethod
+    def pre_save(validated_data: dict) -> dict:
+        # TODO emit events
+        return validated_data
+
+    @staticmethod
+    def post_save(answer: models.Answer) -> models.Answer:
+        # TODO emit events
+        return answer
+
+    @staticmethod
+    @transaction.atomic
+    def create(validated_data, user):
+        if validated_data["question"].type == models.Question.TYPE_FILE:
+            validated_data = __class__._set_file(validated_data)
+
+        if validated_data["question"].type == models.Question.TYPE_TABLE:
+            row_docs = validated_data.pop("row_docs")
+
+        if user:
+            validated_data["created_by_user"] = user
+            validated_data["created_by_group"] = user.group
+
+        answer = models.Answer.objects.create(**validated_data)
+
+        if answer.question.type == models.Question.TYPE_TABLE:
+            answer.create_answer_documents(row_docs)
+
+        return answer
+
+    @staticmethod
+    @transaction.atomic
+    def update(validated_data, answer):
+        if answer.question.type == models.Question.TYPE_TABLE:
+            row_docs = validated_data.pop("row_docs")
+            answer.unlink_unused_rows(docs_to_keep=row_docs)
+
+        if (
+            answer.question.type == models.Question.TYPE_FILE
+            and answer.file.name is not validated_data["file"]
+        ):
+            answer.file.delete()
+            validated_data["file"] = validated_data.pop("value")
+            validated_data = __class__._set_file(validated_data)
+
+        update_model(answer, validated_data)
+
+        if answer.question.type == models.Question.TYPE_TABLE:
+            answer.create_answer_documents(row_docs)
+
+    @staticmethod
+    def _set_file(validated_data):
```

Comment:
This is a "private" method but is used in various other files, doesn't really make sense